README.md
# Vagrant Cloudstack Provider

[![Build Status](https://travis-ci.org/MissionCriticalCloud/vagrant-cloudstack.png?branch=master)](https://travis-ci.org/MissionCriticalCloud/vagrant-cloudstack)
[![Gem Version](https://badge.fury.io/rb/vagrant-cloudstack.png)](http://badge.fury.io/rb/vagrant-cloudstack)
[![Code climate](https://codeclimate.com/github/MissionCriticalCloud/vagrant-cloudstack.png)](https://codeclimate.com/github/MissionCriticalCloud/vagrant-cloudstack)
[![Coverage Status](https://coveralls.io/repos/github/MissionCriticalCloud/vagrant-cloudstack/badge.svg?branch=master)](https://coveralls.io/github/MissionCriticalCloud/vagrant-cloudstack?branch=master)
[![BCH compliance](https://bettercodehub.com/edge/badge/MissionCriticalCloud/vagrant-cloudstack?branch=master)](https://bettercodehub.com/)

This is a fork of [mitchellh AWS Provider](https://github.com/mitchellh/vagrant-aws/).

This is a [Vagrant](http://www.vagrantup.com) 1.5+ plugin that adds a [Cloudstack](http://cloudstack.apache.org)
provider to Vagrant for use with either [Cosmic](https://github.com/MissionCriticalCloud/cosmic) or [Cloudstack](http://cloudstack.apache.org).

## Features

* SSH into the instances.
* Provision the instances with any built-in Vagrant provisioner.
* Minimal synced folder support via `rsync`/`winrm`.

## Usage

Install using standard Vagrant 1.1+ plugin installation methods. After
installing, `vagrant up` and specify the `cloudstack` provider. An example is
shown below.

```
$ vagrant plugin install vagrant-cloudstack
...
$ vagrant up --provider=cloudstack
...
```

## Test status features [Cosmic](https://github.com/MissionCriticalCloud/cosmic) versus [Cloudstack](http://cloudstack.apache.org)
The release of version 1.1.0 - 1.4.0 were (only) released after ALL functional tests passed.
As we, current maintainers, are only using [Cosmic](https://github.com/MissionCriticalCloud/cosmic), we only have availability of a [Cosmic](https://github.com/MissionCriticalCloud/cosmic) cloud, not a CloudStack cloud.

This means that we, maintainers, are not able to actively test the following functionality:
* Basic Zone Network
* Security Groups
* Virtual Router / Isolated network

For any change on these items we kindly request you to run the appropriate functional tests and provide the result in the PR.


## Quick Start

After installing the plugin (instructions above), the quickest way to get
started is to actually make a Vagrantfile that looks like the following, filling in
your information where necessary.

```
Vagrant.configure("2") do |config|
  config.vm.box = "${cloudstack.template_name}"

  config.vm.provider :cloudstack do |cloudstack, override|
    cloudstack.host = "cloudstack.local"
    cloudstack.path = "/client/api"
    cloudstack.port = "8080"
    cloudstack.scheme = "http"
    cloudstack.api_key = "AAAAAAAAAAAAAAAAAAA"
    cloudstack.secret_key = "AAAAAAAAAAAAAAAAAAA"

    cloudstack.service_offering_id = "AAAAAAAAAAAAAAAAAAA"
    cloudstack.disk_offering_id = "AAAAAAAAAAAAAAAAAAA"
    cloudstack.network_id = "AAAAAAAAAAAAAAAAAAA"
    cloudstack.zone_id = "AAAAAAAAAAAAAAAAAAA"
    cloudstack.project_id = "AAAAAAAAAAAAAAAAAAA"
  end
end
```

Or with names instead of ids:

```
Vagrant.configure("2") do |config|
  config.vm.box = "${cloudstack.template_name}"

  config.vm.provider :cloudstack do |cloudstack, override|
    cloudstack.host = "cloudstack.local"
    cloudstack.path = "/client/api"
    cloudstack.port = "8080"
    cloudstack.scheme = "http"
    cloudstack.api_key = "AAAAAAAAAAAAAAAAAAA"
    cloudstack.secret_key = "AAAAAAAAAAAAAAAAAAA"

    cloudstack.service_offering_name = "THE-BESTEST"
    cloudstack.disk_offering_name = "THE-LARGEST-OFFER-AVAILABLE"
    cloudstack.network_name = "WOW-SUCH-FAST-OFFERING"
    cloudstack.zone_name = "District-9"
    cloudstack.name = "doge-is-a-hostname-now"
    # Sadly there is currently no support for the project api in fog.
    cloudstack.project_id = "AAAAAAAAAAAAAAAAAAA"
  end
end
```

And then run `vagrant up --provider=cloudstack`.

This will start an instance in Cloudstack. And assuming your template
on Cloudstack is Vagrant compatible _(vagrant user with official
vagrant pub key in authorized_keys)_ SSH and provisioning will work as
well.

## Configuration

This provider exposes quite a few provider-specific configuration options. Most of the settings
have both an id and a name setting and you can chose to use either (i.e network_id or network_name).
This gives the possibility to use the easier to remember name instead of the UUID,
this will also enable you to upgrade the different settings in your cloud without having
to update UUIDs in your Vagrantfile. If both are specified, the id parameter takes precedence.

* `name` - Hostname of the created machine
* `host` - Cloudstack api host
* `path` - Cloudstack api path
* `port` - Cloudstack api port
* `scheme` - Cloudstack api scheme _(defaults: https (thanks to the resolution order in fog))_
* `api_key` - The api key for accessing Cloudstack
* `secret_key` - The secret key for accessing Cloudstack
* `instance_ready_timeout` - The number of seconds to wait for the instance
  to become "ready" in Cloudstack. Defaults to 120 seconds.
* `domain_id` - Domain id to launch the instance into
* `network_id` - Network uuid(s) that the instance should use
  - `network_id` is single value (e.g. `"AAAA"`) or multiple values (e.g. `["AAAA", "BBBB"]`)
* `network_name` - Network name(s) that the instance should use
  - `network_name` is single value (e.g. `"AAAA"`) or multiple values (e.g. `["AAAA", "BBBB"]`)
* `project_id` - Project uuid that the instance should belong to
* `service_offering_id`- Service offering uuid to use for the instance
* `service_offering_name`- Service offering name to use for the instance
* `template_id` - Template uuid to use for the instance
* `template_name` - Template name to use for the instance, defaults to Vagrants config.vm.box
* `zone_id` - Zone uuid to launch the instance into
* `zone_name` - Zone uuid to launch the instance into
* `keypair` - SSH keypair name, if neither'keypair' nor 'ssh_key' have been specified, a temporary keypair will be created
* `static_nat` - static nat for the virtual machine
* `pf_ip_address_id` - IP address ID for port forwarding rule
* `pf_ip_address` - IP address for port forwarding rule
* `pf_public_port` - Public Communicator port for port forwarding rule
* `pf_public_rdp_port` - Public RDP port for port forwarding rule
* `pf_public_port_randomrange` - If public port is omited, a port from this range wll be used (default `{:start=>49152, :end=>65535}`)
* `pf_private_port` - Private port for port forwarding rule (defaults to respective Communicator protocol)
* `pf_open_firewall` - Flag to enable/disable automatic open firewall rule (by CloudStack)
* `pf_trusted_networks` - Array of CIDRs or (array of) comma-separated string of CIDRs to network(s) to 
  - automatically (by plugin) generate firewall rules for, ignored if `pf_open_firewall` set `true`
  - use as default for firewall rules where source CIDR is missing
* `port_forwarding_rules` - Port forwarding rules for the virtual machine
* `firewall_rules` - Firewall rules
* `display_name` - Display name for the instance
* `group` - Group for the instance
* `ssh_key` - Path to a private key to be used with ssh _(defaults to Vagrant's `config.ssh.private_key_path`)_
* `ssh_user` - User name to be used with ssh _(defaults to Vagrant's `config.ssh.username`)_
* `ssh_network_id` - The network_id to be used when loging in to the vm via ssh _(defaults to first nic)_
* `ssh_network_name` - The network_name to be used when loging in to the vm via ssh _(defaults to first nic)_
  - Use either `ssh_network_id` or `ssh_network_name`. If specified both , use `ssh_network_id`
* `vm_user` - User name to be used with winrm _(defaults to Vagrant's `config.winrm.username`)_
* `vm_password` - Password to be used with winrm. _(If the CloudStack template is "Password Enabled", leaving this unset will trigger the plugin to retrieve and use it.)_
* `private_ip_address` - private (static)ip address to be used by the virtual machine
* `expunge_on_destroy` - Flag to enable/disable expunge vm on destroy

These can be set like typical provider-specific configuration:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|
    cloudstack.api_key = "foo"
    cloudstack.secret_key = "bar"
  end
end
```

In addition to the above top-level configs, you can use the `region_config`
method to specify region-specific overrides within your Vagrantfile. Note
that the top-level `region` config must always be specified to choose which
region you want to actually use, however. This looks like this:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|
    cloudstack.api_key = "foo"
    cloudstack.secret_key = "bar"
    cloudstack.domain = "internal"

    # Simple domain config
    cloudstack.domain_config "internal", :network_id => "AAAAAAAAAAAAAAAAAAA"

    # More comprehensive region config
    cloudstack.domain_config "internal" do |domain|
      domain.network_id = "AAAAAAAAAAAAAAAAAAA"
      domain.service_offering_id = "AAAAAAAAAAAAAAAAAAA"
    end
  end
end
```

The domain-specific configurations will override the top-level
configurations when that domain is used. They otherwise inherit
the top-level configurations, as you would probably expect.

## Networks

Networking features in the form of `config.vm.network` are not
supported with `vagrant-cloudstack`, currently. If any of these are
specified, Vagrant will emit a warning, but will otherwise boot
the Cloudstack machine.

### Basic networking versus Advanced networking

The plugin will determine this network type dynamically from the zone.
The setting `network_type` in the Vagrant file has been deprecated,
and is silently ignored.

### Basic Networking

If the network type of your zone is `basic`, you can use Security
Groups and associate rules in your Vagrantfile.

If you already have Security Groups, you can associate them to your
instance, with their IDs:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|
    cloudstack.api_key = "foo"
    cloudstack.secret_key = "bar"
    cloudstack.security_group_ids = ['aaaa-bbbb-cccc-dddd', '1111-2222-3333-4444']
  end
end
```

or their names:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|
    cloudstack.api_key = "foo"
    cloudstack.secret_key = "bar"
    cloudstack.security_group_names = ['
min_fantastiska_security_group', 'another_security_grupp']
  end
end
```

But you can also create your Security Groups in the Vagrantfile:

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|
    cloudstack.api_key = "foo"
    cloudstack.secret_key = "bar"
    cloudstack.security_groups = [
      {
        :name         => "Awesome_security_group",
        :description  => "Created from the Vagrantfile",
          :rules => [
        {:type => "ingress", :protocol => "TCP", :startport => 22, :endport => 22, :cidrlist => "0.0.0.0/0"},
        {:type => "ingress", :protocol => "TCP", :startport => 80, :endport => 80, :cidrlist => "0.0.0.0/0"},
        {:type => "egress",  :protocol => "TCP", :startport => 81, :endport => 82, :cidrlist => "1.2.3.4/24"},
    ]
      }
    ]
  end
end
```

### Static NAT, Firewall, Port forwarding

You can create your static nat, firewall and port forwarding rules in the Vagrantfile.
You can use this rule to access virtual machine from an external machine.

The rules created in Vagrantfile are removed when the virtual machine is destroyed.

```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|

    override.ssh.host = "X.X.X.X"

    cloudstack.static_nat = [
      { :ipaddress => "A.A.A.A"}
    ]

    cloudstack.port_forwarding_rules = [
      { :ipaddress => "X.X.X.X", :protocol => "tcp", :publicport => 22, :privateport  => 22, :openfirewall => false },
      { :ipaddress => "X.X.X.X", :protocol => "tcp", :publicport => 80, :privateport  => 80, :openfirewall => false }
    ]

    cloudstack.firewall_rules = [
      { :ipaddress => "A.A.A.A", :cidrlist  => "1.2.3.4/24", :protocol => "icmp", :icmptype => 8, :icmpcode => 0 },
      { :ipaddress => "X.X.X.X", :cidrlist  => "1.2.3.4/24", :protocol => "tcp", :startport => 22, :endport => 22 },
      { :ipaddress => "X.X.X.X", :cidrlist  => "1.2.3.4/24", :protocol => "tcp", :startport => 80, :endport => 80 }
    ]

  end
end
```
Most values in the firewall and portforwarding rules are not mandatory, except either startport/endport or privateport/publicport
* `:ipaddress` - defaults to `pf_ip_address`
* `:protocol` - defaults to `'tcp'`
* `:publicport` - defaults to `:privateport`
* `:privateport` - defaults to `:publicport`
* `:openfirewall` - defaults to `pf_open_firewall`
* `:cidrlist` - defaults to `pf_trusted_networks`
* `:startport` - defaults to `:endport`
* `:endport` - not required by CloudStack


For only allowing Vagrant to access the box for further provisioning (SSH/WinRM), and opening the Firewall for some subnets, the following config is sufficient:
```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|
    cloudstack.pf_open_firewall      = "false"
    cloudstack.pf_ip_address         = X.X.X.X
    cloudstack.pf_trusted_networks   = [ "1.2.3.4/24" , "11.22.33.44/32" ]
  end
end
```
Where X.X.X.X is the ip of the respective CloudStack network, this will automatically map the port of the used Communicator (SSH/Winrm) via a random public port, open the Firewall and set Vagrant to use it.

The plugin can also automatically generate firewall rules off of the portforwarding rules:
```ruby
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :cloudstack do |cloudstack|

    cloudstack.pf_trusted_networks   = "1.2.3.4/24,11.22.33.44/32"
    cloudstack.port_forwarding_rules = [
      { :privateport  => 22, :generate_firewall => true },
      { :privateport  => 80, :generate_firewall => true }
    ]

  end
end
```

### Virtual Router versus VPC
Both Virtual Routers and VPCs are supported when using port-forwarding and firewall. This is automatically determined by the specific `pf_ip_address`.

Note that there are architectural differences in CloudStack which the configuration must adhere to.

For VPC:
* `pf_open_firewall` will be ignored as global setting and (specifically) in `port_forwarding_rules`
* for `firewall_rules` to open access for `port_forwarding_rules`, the firewall rule should allow traffic for the `:privateport` port.

For Virtual Router:
* for `firewall_rules` to open access for `port_forwarding_rules`, the firewall rule should allow traffic for the `:publicport` port.

Usage of other attributes and features work with both network types. Such as `:generate_firewall` for portforwarding rules, or `pf_trusted_networks` to automatically generate rules for the Communicator.

## Synced Folders

There is minimal support for synced folders. Upon `vagrant up`, `vagrant reload`, and `vagrant provision`, the Cloudstack provider will use `rsync` (if available) to uni-directionally sync the folder to the remote machine over SSH, and use Vagrant plugin `vagrant-winrm-syncedfolders` (if available) to uni-directionally sync the folder to the remote machine over WinRM.

This is good enough for all built-in Vagrant provisioners (shell,

CHANGELOG.md
# 1.5.2 (02 Oct 2019)

Fixed bugs:

- `store_passwd` would sometimes try to lookup an async job status without including the job id causing an exception being raised [\#187](https://github.com/MissionCriticalCloud/vagrant-cloudstack/issues/187)

# 1.5.1 (18 Apr 2019)

**Implemented enhancements:**

- Sync resources by name: sometimes a resource (e.g. templates) were not correctly found; resources are now looked up by specifying the name of the resource

# 1.5.0 (17 Oct 2018)

**Implemented enhancements:**

- Provide human readable error when `sync_resource` cannot sync name-\>id [\#163](https://github.com/MissionCriticalCloud/vagrant-cloudstack/issues/163)
- Deduplicate code in read\_\[ssh|winrm|rdp\]\_info [\#150](https://github.com/MissionCriticalCloud/vagrant-cloudstack/issues/150)
- CloudStack acquire Public IP address [\#87](https://github.com/MissionCriticalCloud/vagrant-cloudstack/issues/87)

**Fixed bugs:**

- Make vagrant-cloudstack able to run in \(Docker\) container [\#157](https://github.com/MissionCriticalCloud/vagrant-cloudstack/issues/157)

**Closed issues:**

- Write Code Once [\#172](https://github.com/MissionCriticalCloud/vagrant-cloudstack/issues/172)
- Template name used is mis-reported [\#56](https://github.com/MissionCriticalCloud/vagrant-cloudstack/issues/56)

**Merged pull requests:**

- Fix the travis setup. [\#179](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/179) ([greut](https://github.com/greut))
- Update Vagrant to 2.0.3, unbork bundler. [\#178](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/178) ([greut](https://github.com/greut))
- Add testing for different scenarios for basic zone implementation [\#176](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/176) ([bheuvel](https://github.com/bheuvel))
- Refactoring of run\_instance, config and added testing [\#175](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/175) ([bheuvel](https://github.com/bheuvel))
- Unit testing covering a \(most\) basic VM deployment [\#174](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/174) ([bheuvel](https://github.com/bheuvel))
- Refactor read info classes using a super class [\#173](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/173) ([bheuvel](https://github.com/bheuvel))
- Improve error message/handling [\#170](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/170) ([bheuvel](https://github.com/bheuvel))
- Don't set network\_ids when using basic networking [\#168](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/168) ([hrak](https://github.com/hrak))
- Pin kitchen-vagrant and vagrant-winrm [\#167](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/167) ([bheuvel](https://github.com/bheuvel))
- Fix user dependency [\#165](https://github.com/MissionCriticalCloud/vagrant-cloudstack/pull/165) ([bheuvel](https://github.com/bheuvel))

# 1.4.0 (Nov 30, 2016)
* Support multiple network ids and network names (PR #148)
* Add ssh_network_id and ssh_network_name configuration (PR #149)
* Add (firewall management for) VPC support (PR #151)
* Remove additional data disk on destroy (PR #152)
* Add Docker containers for development and testing (PR #159)
* CloudStack >= 4.6 list all offerings / templates. (PR #161)

# 1.3.0 (Mar 24, 2016)
* Automate port forwarding for RDP for Windows guests (PR #117)
* Specify `trusted_networks` by Array (PR #121)
* Random public port range is specified by hash using `:start` and `:end` (PR #122)
* Generate firewall rule from port forward rules automatically (PR #123)
* Let firewall rule elements (`ipaddress`,`protocol`,`startport`) use defaults (PR #123)
* Generate SSH keypair for lifetime of VM (PR #125)
* Deprecate usage in Vagrantfile of `cloudstack.network_type` (PR #134)
* Determine dynamically from ZONE if 'Basic' or 'Advanced' networking  (PR #134)
* Allow trusted_networks as both string or array (PR #146)

# 1.2.0 (Sep 4, 2015)
* Add support for disk offering (PR#89)
* Fix bug open file handles on Windows (PR#98)
* Add support for Windows guests (PR#96)
* Automate port forwarding for the Communicator (PR#104)
* Allow setting to open Firewall to specific network (PR#99)
* Make config.vm.box an optional setting in Vagrantfile (PR#105)

# 1.1.0 (May 26, 2015)
* Allow setting VM private IP in config (PR #73)
* Fix several coding style issues (PR #76, #77, #78)
* Fix bug when destroying VM created with Static NAT and firewall rules (PR #81)
* Allow expunging VM on destroy (PR #75)
* Make `network_type` optional, and defaulting to "Basic" (PR #82)

# 1.0.0 (May 5, 2015)
* Use vagrant's built-in rsync synced folder (PR #57)
* Enable creating custom static NAT, port forwarding, firewall rules (PR #59)
* Fixed bug when `network_id` and `network_name` are not specified in Vagrantfile (PR #59)
* Enable setting SSH user name and private key to access VM (PR #64)
* Fixed bug when `vagrant destroy` destroys other VMs (PR #66)
* Enable toggling port forwarding automatically adding an open firewall rule (PR #70)

# 0.10.0 (Sep 11, 2014)
* Clean up code base DRY
* Improve documentation
* Use URL safe base 64 encoding

# 0.9.1 (Jun 29, 2014)
* Fixed a bug intrcduced in 0.9.0, where we failed to fetch correct
  name because we didn't pass the correct parameters.

# 0.9.0 (Jun 25, 2014)
* Clean up of dead code and comments.
* Re-organize imports and code.
* Corrects the dependency on Vagrant 1.5+
* Updates documentation to refelect the above.
* Now supports setting the machine hostname in Vagrantfile

# 0.8.0 (Jun 24, 2014)
* Remove unused code.
* Add support for specifying most resources by name. Where applicable there is
  a matching _name config for the _id configs.
  Note: in this release there were no support for the project apis in fog,
  therefore there is no project_name config.

# 0.7.0 (Jun 17, 2014)
* Change the resolution order of how we discover the scheme to talk to the cloud.
  This is a possibly breaking change.


lib/vagrant-cloudstack/config.rb
require "vagrant"

module VagrantPlugins
  module Cloudstack
    class Config < Vagrant.plugin("2", :config)
      INSTANCE_VAR_DEFAULT_NIL = %w(host name path port domain_id network_id network_name project_id service_offering_id service_offering_name
           template_id template_name zone_id zone_name keypair pf_ip_address_id pf_ip_address pf_public_port
           pf_public_rdp_port pf_private_port pf_trusted_networks display_name group user_data ssh_key ssh_user
           ssh_network_id ssh_network_name vm_user vm_password private_ip_address affinity_group_ids affinity_group_names).freeze
      INSTANCE_VAR_DEFAULT_EMPTY_ARRAY = %w(static_nat port_forwarding_rules firewall_rules security_group_ids security_group_names security_groups).freeze

      # Cloudstack api host.
      #
      # @return [String]
      attr_accessor :host

      # Hostname for the machine instance
      # This will be passed through to the api.
      #
      # @return [String]
      attr_accessor :name

      # Cloudstack api path.
      #
      # @return [String]
      attr_accessor :path

      # Cloudstack api port.
      #
      # @return [String]
      attr_accessor :port

      # Cloudstack api scheme
      #
      # @return [String]
      attr_accessor :scheme

      # The API key for accessing Cloudstack.
      #
      # @return [String]
      attr_accessor :api_key

      # The secret key for accessing Cloudstack.
      #
      # @return [String]
      attr_accessor :secret_key

      # The timeout to wait for an instance to become ready.
      #
      # @return [Fixnum]
      attr_accessor :instance_ready_timeout

      # Domain id to launch the instance into.
      #
      # @return [String]
      attr_accessor :domain_id

      # Network uuid(s) that the instance should use
      #
      # @return [String,Array]
      attr_accessor :network_id

      # Network name(s) that the instance should use
      #
      # @return [String,Array]
      attr_accessor :network_name

      # Network Type
      #
      # @return [String]
      attr_accessor :network_type

      # Project uuid that the instance should belong to
      #
      # @return [String]
      attr_accessor :project_id

      # Service offering uuid to use for the instance
      #
      # @return [String]
      attr_accessor :service_offering_id

      # Service offering name to use for the instance
      #
      # @return [String]
      attr_accessor :service_offering_name

      # Disk offering uuid to use for the instance
      #
      # @return [String]
      attr_accessor :disk_offering_id

      # Disk offering name to use for the instance
      #
      # @return [String]
      attr_accessor :disk_offering_name

      # Template uuid to use for the instance
      #
      # @return [String]
      attr_accessor :template_id

      # Template name to use for the instance
      #
      # @return [String]
      attr_accessor :template_name

      # Zone uuid to launch the instance into. If nil, it will
      # launch in default project.
      #
      # @return [String]
      attr_accessor :zone_id

      # Zone name to launch the instance into. If nil, it will
      # launch in default project.
      #
      # @return [String]
      attr_accessor :zone_name

      # The name of the keypair to use.
      #
      # @return [String]
      attr_accessor :keypair

      # Paramters for Static NAT
      #
      # @return [String]
      attr_accessor :static_nat

      # IP address id to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_ip_address_id

      # IP address to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_ip_address

      # public port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_public_port

      # public port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_public_rdp_port

      # private port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_private_rdp_port

      # public port to use for port forwarding rule
      #
      # @return [Range]
      attr_accessor :pf_public_port_randomrange

      # private port to use for port forwarding rule
      #
      # @return [String]
      attr_accessor :pf_private_port

      # flag to enable/disable automatic open firewall rule
      #
      # @return [Boolean]
      attr_accessor :pf_open_firewall

      # CIDR List string of trusted networks
      #
      # @return [String]
      attr_accessor :pf_trusted_networks

      # comma separated list of port forwarding rules
      # (hash with rule parameters)
      #
      # @return [Array]
      attr_accessor :port_forwarding_rules

      # comma separated list of firewall rules
      # (hash with rule parameters)
      #
      # @return [Array]
      attr_accessor :firewall_rules

      # comma separated list of security groups id that going
      # to be applied to the virtual machine.
      #
      # @return [Array]
      attr_accessor :security_group_ids

      # comma separated list of security groups name that going
      # to be applied to the virtual machine.
      #
      # @return [Array]
      attr_accessor :security_group_names

      # comma separated list of security groups
      # (hash with ingress/egress rules)
      # to be applied to the virtual machine.
      #
      # @return [Array]
      attr_accessor :security_groups

      # display name for the instance
      #
      # @return [String]
      attr_accessor :display_name

      # group for the instance
      #
      # @return [String]
      attr_accessor :group

      # The user data string
      #
      # @return [String]
      attr_accessor :user_data

      # The key to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_key

      # The username to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_user

      # The network_id to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_network_id

      # The network_name to be used when loging in to the vm via ssh
      #
      # @return [String]
      attr_accessor :ssh_network_name

      # The username to be used when loging in to the vm
      #
      # @return [String]
      attr_accessor :vm_user

      # The username to be used when loging in to the vm
      #
      # @return [String]
      attr_accessor :vm_password

      # Private ip for the instance
      #
      # @return [String]
      attr_accessor :private_ip_address

      # Affinity Group IDs for the instance
      #
      # @return [String]
      attr_accessor :affinity_group_ids

      # Affinity Group Names for the instance
      #
      # @return [String]
      attr_accessor :affinity_group_names

      # flag to enable/disable expunge vm on destroy
      #
      # @return [Boolean]
      attr_accessor :expunge_on_destroy

      def initialize(domain_specific = false)
        # Initialize groups in bulk, re-use these groups to set defaults in bulk
        INSTANCE_VAR_DEFAULT_NIL.each do |instance_variable|
          instance_variable_set("@#{instance_variable}", UNSET_VALUE)
        end
        # Initialize groups in bulk, re-use these groups to set defaults in bulk
        INSTANCE_VAR_DEFAULT_EMPTY_ARRAY.each do |instance_variable|
          instance_variable_set("@#{instance_variable}", UNSET_VALUE)
        end

        @scheme                     = UNSET_VALUE
        @api_key                    = UNSET_VALUE
        @secret_key                 = UNSET_VALUE
        @instance_ready_timeout     = UNSET_VALUE
        @network_type               = UNSET_VALUE
        @pf_private_rdp_port        = UNSET_VALUE
        @pf_public_port_randomrange = UNSET_VALUE
        @pf_open_firewall           = UNSET_VALUE
        @expunge_on_destroy         = UNSET_VALUE

        # Internal state (prefix with __ so they aren't automatically
        # merged)
        @__compiled_domain_configs = {}
        @__finalized               = false
        @__domain_config           = {}
        @__domain_specific         = domain_specific
      end

      # Allows domain-specific overrides of any of the settings on this
      # configuration object. This allows the user to override things like
      # template and keypair name for domains. Example:
      #
      #     cloudstack.domain_config "abcd-ef01-2345-6789" do |domain|
      #       domain.template_id = "1234-5678-90ab-cdef"
      #       domain.keypair_name = "company-east"
      #     end
      #
      # @param [String] domain The Domain name to configure.
      # @param [Hash] attributes Direct attributes to set on the configuration
      #   as a shortcut instead of specifying a full block.
      # @yield [config] Yields a new domain configuration.
      def domain_config(domain, attributes=nil, &block)
        # Append the block to the list of domain configs for that domain.
        # We'll evaluate these upon finalization.
        @__domain_config[domain] ||= []

        # Append a block that sets attributes if we got one
        if attributes
          attr_block = lambda do |config|
            config.set_options(attributes)
          end

          @__domain_config[domain] << attr_block
        end

        # Append a block if we got one
        @__domain_config[domain] << block if block_given?
      end

      #-------------------------------------------------------------------
      # Internal methods.
      #-------------------------------------------------------------------

      def merge(other)
        super.tap do |result|
          # Copy over the domain specific flag. "True" is retained if either
          # has it.
          new_domain_specific = other.instance_variable_get(:@__domain_specific)
          result.instance_variable_set(
              :@__domain_specific, new_domain_specific || @__domain_specific)

          # Go through all the domain configs and prepend ours onto
          # theirs.
          new_domain_config = other.instance_variable_get(:@__domain_config)
          @__domain_config.each do |key, value|
            new_domain_config[key] ||= []
            new_domain_config[key] = value + new_domain_config[key]
          end

          # Set it
          result.instance_variable_set(:@__domain_config, new_domain_config)

          # Merge in the tags
          result.tags.merge!(self.tags)
          result.tags.merge!(other.tags)
        end
      end

      def finalize!
        INSTANCE_VAR_DEFAULT_NIL.each do |instance_variable|
          # ... must be nil, since we can't default that
          instance_variable_set("@#{instance_variable}", nil) if
              instance_variable_get("@#{instance_variable}") == UNSET_VALUE
        end

        INSTANCE_VAR_DEFAULT_EMPTY_ARRAY.each do |instance_variable|
          # ... must be empty array
          instance_variable_set("@#{instance_variable}", []) if
              instance_variable_get("@#{instance_variable}") == UNSET_VALUE
        end

        # We default the scheme to whatever the user has specifid in the .fog file
        # *OR* whatever is default for the provider in the fog library
        @scheme                 = nil if @scheme == UNSET_VALUE

        # Try to get access keys from environment variables, they will
        # default to nil if the environment variables are not present
        @api_key                = ENV['CLOUDSTACK_API_KEY'] if @api_key == UNSET_VALUE
        @secret_key             = ENV['CLOUDSTACK_SECRET_KEY'] if @secret_key == UNSET_VALUE

        # Set the default timeout for waiting for an instance to be ready
        @instance_ready_timeout = 120 if @instance_ready_timeout == UNSET_VALUE

        # NetworkType is 'Advanced' by default
        @network_type           = "Advanced" if @network_type == UNSET_VALUE

        # Private rdp port defaults to 3389
        @pf_private_rdp_port     = 3389 if @pf_private_rdp_port == UNSET_VALUE

        # Public port random-range, default to rfc6335 'Dynamic Ports'; "(never assigned)"
        @pf_public_port_randomrange = {:start=>49152, :end=>65535} if @pf_public_port_randomrange == UNSET_VALUE

        # Open firewall is true by default (for backwards compatibility)
        @pf_open_firewall       = true if @pf_open_firewall == UNSET_VALUE

        # expunge on destroy is nil by default
        @expunge_on_destroy     = false if @expunge_on_destroy == UNSET_VALUE

        # Compile our domain specific configurations only within
        # NON-DOMAIN-SPECIFIC configurations.
        unless @__domain_specific
          @__domain_config.each do |domain, blocks|
            config = self.class.new(true).merge(self)

            # Execute the configuration for each block
            blocks.each { |b| b.call(config) }

            # The domain name of the configuration always equals the domain config name:
            config.domain = domain

            # Finalize the configuration
            config.finalize!

            # Store it for retrieval
            @__compiled_domain_configs[domain] = config
          end
        end

        # Mark that we finalized
        @__finalized = true
      end

      def validate(machine)
        errors = []

        if @domain
          # Get the configuration for the domain we're using and validate only that domain.
          config = get_domain_config(@domain)

          unless config.use_fog_profile
            errors << I18n.t("vagrant_cloudstack.config.api_key_required") if \
               config.access_key_id.nil?
            errors << I18n.t("vagrant_cloudstack.config.secret_key_required") if \
               config.secret_access_key.nil?
          end
        end

        {"Cloudstack Provider" => errors}
      end

      # This gets the configuration for a specific domain. It shouldn't
      # be called by the general public and is only used internally.
      def get_domain_config(name)
        raise 'Configuration must be finalized before calling this method.' unless @__finalized

        # Return the compiled domain config
        @__compiled_domain_configs[name] || self
      end
    end
  end
end
